polaris-migrator/src/utilities/sass.ts

```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
+        ) => false | void,
+      ) => void;
+      walkRules: <T extends Container>(
+        root: T,
+        ruleWalker: (decl: PostCSSRule) => false | void,
+      ) => void;
+    };
+  },
+  context: PluginContext,
+) => (root: Root) => void;
+
+// Expose a stylelint-like API for creating sass migrators so we can easily
+// migrate to that tool in the future.
+function convertStylelintRuleToPostcssProcessor(ruleFn: StylelintRule) {
+  return (fileInfo: FileInfo, _: API, options: Options) => {
+    const plugin: Plugin = {
+      postcssPlugin: ruleFn.ruleName,
+      // PostCSS will rewalk the AST every time a declaration/rule/etc is
+      // mutated by a plugin. This can be useful in some cases, but in ours we
+      // only want single-pass behaviour.
+      //
+      // This can be avoided in 2 ways:
+      //
+      // 1) Flagging each declaration as we pass it, then skipping it on
+      //    subsequent passes.
+      // 2) Using postcss's Once() plugin callback.
+      //
+      // stylelint also uses `Once()`, so we're able to remove this once we've
+      // migrated:
+      // https://github.com/stylelint/stylelint/blob/cb425cb/lib/postcssPlugin.js#L22
+      Once(root: Root) {
+        // NOTE: For fullest compatibility with stylelint, we initialise the
+        // rule here _inside_ the postcss Once function just like stylelint
+        // does. I _think_ this means multiple passes can be performed without
+        // rules accidentally retaining scoped variables, etc.
+        const ruleProcessor = ruleFn(
+          // Normally, this comes from stylelint config, but for this shim we
+          // just hard-code it, and instead rely on the "seconary" options
+          // object for passing through the jscodeshift options.
+          true,
+          {
+            ...options,
+            partialFixStrategy: 'report',
+            __injectReportsAsComments: true,
+          },
+          {fix: true},
+        );
+
+        ruleProcessor(root);
+      },
+    };
+
+    return postcss(plugin).process(fileInfo.source, {
+      syntax: require('postcss-scss'),
+    }).css;
+  };
+}
+
+export function createStylelintRule(ruleName: string, ruleFn: PolarisMigrator) {
+  const wrappedRule: StylelintRule = ((
+    primary,
+    {
+      partialFixStrategy = 'fix',
+      __injectReportsAsComments = false,
+      ...secondaryOptions
+    }: PluginOptions,
+    context,
+  ) => {
+    const reports = new Map<PostCSSNode, Report[]>();
+    const suggestions = new Map<PostCSSNode, string>();
+
+    const addDedupedReport = (newReport: Report) => {
+      if (!reports.has(newReport.node)) {
+        reports.set(newReport.node, []);
+      }
+
+      const reportsForNode = reports.get(newReport.node)!;
+
+      if (
+        reportsForNode.findIndex(
+          (existingReport) =>
+            existingReport.severity === newReport.severity &&
+            existingReport.message === newReport.message,
+        ) === -1
```

Comment:
Deduping reports to avoid noise. Particularly useful for example `padding: 10px 10px 10px 10px` - we only want 1 warning, not 4 identical ones.